### PR TITLE
Fix build errors for WindowsCE

### PR DESCRIPTION
**Pull request by aSoujyuTanaka**
I have confirmed that the build libraries works with our customized cURL based code. It can show a page for TLS1.3 only. Build instruction is below.
https://qiita.com/souju/items/94117c024862f57459c3

##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: crypto/bio/bss_file.c [hunk 1]
```diff
@@ -33,8 +33,6 @@
 # include "bio_lcl.h"
 # include <openssl/err.h>
 
-# if !defined(OPENSSL_NO_STDIO)
-
```
Comment (mattcaswell) [inline crypto/bio/bss_file.c]: The code below uses stdio capabilities so removing this guard seems odd and liable to break things.

File: crypto/dso/dso_win32.c [hunk 2]
```diff
@@ -500,6 +501,106 @@ typedef HANDLE(WINAPI *CREATETOOLHELP32SNAPSHOT) (DWORD, DWORD);
...
+    if (!(*module_first) (hModuleSnap, &me32)) {
+        (*close_snap) (hModuleSnap);
+        FreeLibrary(dll);
+        DSOerr(DSO_F_WIN32_PATHBYADDR, DSO_R_FAILURE);
+        return -1;
+    }
+
+    do {
+        if ((BYTE *) addr >= me32.modBaseAddr &&
+            (BYTE *) addr < me32.modBaseAddr + me32.modBaseSize) {
```
Comment (mattcaswell) [inline crypto/dso/dso_win32.c]: This is undefined behaviour not allowed by the C spec. You are only allowed to compare pointers if they both point into the same array.

File: crypto/dso/dso_win32.c [hunk 3]
```diff
@@ -500,6 +501,106 @@ typedef HANDLE(WINAPI *CREATETOOLHELP32SNAPSHOT) (DWORD, DWORD);
...
+# endif
+    module_first = (MODULE32) GetProcAddress(dll, "Module32First");
+    module_next = (MODULE32) GetProcAddress(dll, "Module32Next");
+
+    hModuleSnap = (*create_snap) (TH32CS_SNAPMODULE, 0);
+    if (hModuleSnap == INVALID_HANDLE_VALUE) {
+        FreeLibrary(dll);
+        DSOerr(DSO_F_WIN32_PATHBYADDR, DSO_R_UNSUPPORTED);
+        return -1;
+    }
```
Comment (levitte) [inline crypto/dso/dso_win32.c]: Everything from line 523 are things I do not understand. Mind commenting a bit?

File: crypto/dso/dso_win32.c (hunk 2 above)
Comment (levitte) [reply crypto/dso/dso_win32.c]: We have to remember, though, that this is the Windows realm...

Comment (aSoujyuTanaka) [reply crypto/dso/dso_win32.c]: I have changed to casting in unsigned numeric value before comparison.

File: crypto/bio/bss_file.c (hunk 1 above)
Comment (aSoujyuTanaka) [reply crypto/bio/bss_file.c]: I reverted this change.

File: crypto/dso/dso_win32.c (hunk 3 above)
Comment (aSoujyuTanaka) [reply crypto/dso/dso_win32.c]: I added comment.

File: include/internal/refcount.h [hunk 4]
```diff
@@ -100,17 +100,29 @@ static __inline int CRYPTO_DOWN_REF(volatile int *val, int *ret, void *lock)
...
+#     pragma intrinsic(_InterlockedExchangeAdd)
+#     define INTERLOCKED_EXCHANGE_ADD _InterlockedExchangeAdd
+#    else
+#     if _WIN32_WCE >= 0x600
+       extern long __cdecl _InterlockedExchangeAdd(long volatile*, long);
+#      define INTERLOCKED_EXCHANGE_ADD _InterlockedExchangeAdd
+#     else
+       // under Windows CE we still have old-style Interlocked* functions
+       extern long __cdecl InterlockedExchangeAdd(long volatile*, long);
+#      define INTERLOCKED_EXCHANGE_ADD InterlockedExchangeAdd
```
Comment (mspncp) [inline include/internal/refcount.h]: I would prefer if you would have only a single `#define` here (and remove the other two) and revert the changes in `CRYPTO_UP_REF()` and `CRYPTO_DOWN_REF()`.
```
#      define _InterlockedExchangeAdd InterlockedExchangeAdd
```

**Comment by mattcaswell** [comment]
close/reopen to kick CLA bot.

File: crypto/dso/dso_win32.c [hunk 5]
```diff
@@ -500,6 +501,111 @@ typedef HANDLE(WINAPI *CREATETOOLHELP32SNAPSHOT) (DWORD, DWORD);
...
+#  else
+            {
+                int i, len = (int)wcslen(me32.szExePath);
+                if (sz <= 0)
+                    return len + 1;
+                if (len >= sz)
+                    len = sz - 1;
+                for (i = 0; i < len; i++)
+                    path[i] = (char)me32.szExePath[i];
+                path[len++] = 0;
```
Comment (mattcaswell) [inline crypto/dso/dso_win32.c]: Perhaps `'\0'` is better than 0 here? And below.

Comment (aSoujyuTanaka) [reply crypto/dso/dso_win32.c]: Yes, it is better using '\0' at end of string.

**Comment by aSoujyuTanaka** [comment]
Thank you.

**Comment by mattcaswell** [comment]
Pushed to master and 1.1.1. I squashed the various fixup commits.
